lib/api/objectPutTagging.js

```diff
@@ -0,0 +1,95 @@
+import aclUtils from '../utilities/aclUtils';
+
+import async from 'async';
+import { errors } from 'arsenal';
+
+import { decodeVersionId, getVersionIdResHeader }
+  from './apiUtils/object/versioning';
+
+import { metadataValidateBucketAndObj } from '../metadata/metadataUtils';
+import { pushMetric } from '../utapi/utilities';
+import collectCorsHeaders from '../utilities/collectCorsHeaders';
+import metadata from '../metadata/wrapper';
+import { parseTagXml } from './apiUtils/object/tagging';
+
+/**
+ * Object Put Tagging - Adds tag(s) to object
+ * @param {AuthInfo} authInfo - Instance of AuthInfo class with requester's info
+ * @param {object} request - http request object
+ * @param {object} log - Werelogs logger
+ * @param {function} callback - callback to server
+ * @return {undefined}
+ */
+export default function objectPutTagging(authInfo, request, log, callback) {
+    log.debug('processing request', { method: 'objectPutTagging' });
+
+    if (!aclUtils.checkGrantHeaderValidity(request.headers)) {
+        log.trace('invalid acl header');
+        return callback(errors.InvalidArgument);
+    }
+
+    const bucketName = request.bucketName;
+    const objectKey = request.objectKey;
+
+    const decodedVidResult = decodeVersionId(request.query);
+    if (decodedVidResult instanceof Error) {
+        log.trace('invalid versionId query', {
+            versionId: request.query.versionId,
+            error: decodedVidResult,
+        });
+        return callback(decodedVidResult);
+    }
+    const reqVersionId = decodedVidResult;
+
+    const metadataValParams = {
+        authInfo,
+        bucketName,
+        objectKey,
+        requestType: 'bucketOwnerAction',
+        versionId: reqVersionId,
+    };
+
+    return async.waterfall([
+        next => metadataValidateBucketAndObj(metadataValParams, log,
+          (err, bucket, objectMD) => next(err, bucket, objectMD)),
+        (bucket, objectMD, next) => {
+            log.trace('parsing tag(s)');
+            return parseTagXml(request.post, log, (err, tags) => {
+                if (err) {
+                    return next(err);
+                }
+                if (!objectMD) {
+                    const err = reqVersionId ? errors.NoSuchVersion :
+                        errors.NoSuchKey;
+                    return next(err, bucket);
+                }
```

Comment:
Fixed. 

I decided to put "putObjectTagging" versioning tests to a separate file in the test/versioning folder to make it clearer and consistent.
I tested against real AWS and both putting tag with AND without version id if version specified is a delete marker should return `MethodNotAllowed`

Thanks for the review :)